### PR TITLE
Environment variable registration, deregistration (addresses #3227)

### PR DESCRIPTION
**Pull request by dotsdl**
This PR addresses #3227.

Added `xonsh.environ.Env.register` and `xonsh.environ.Env.deregister` methods for registration, deregistration of new environment variables with types, default values, and docstrings. Currently registration/deregistration is not exposed to the user, so the addition of these gives the ability to add new variables with behavior like e.g. `PATH`.

### CODE REVIEW

**Comment by dotsdl** [comment]
Needs at least:
1. explicit tests
2. recommendations for the way we expose the variable machinery in these functions.

File: xonsh/environ.py [hunk 1]
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
...
+        defaultval : optional
+            Default value for variable. ``ValueError`` raised if type does not match
+            that specified by `vartype` (or `ens_validate`).
+        vardoc : str, optional
+            Docstring for variable.
+        ens_validate : func, optional
+            Function to validate type.
+        ens_convert : func, optional
+            Function to convert variable from a string representation to its type.
+        ens_detype : func, optional
```
Comment (scopatz) [inline xonsh/environ.py]: These could probably drop the `ens_` prefix

File: xonsh/environ.py [hunk 2]
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
...
+        ----------
+        varname : str
+            Environment variable name to register. Typically all caps.
+        vartype : str, optional,  {'bool', 'str', 'path', 'int', 'float'}
+            Variable type. If not one of the available presets, use `ens_validate`,
+            `ens_convert`, and `ens_detype` to specify type behavior.
+        defaultval : optional
+            Default value for variable. ``ValueError`` raised if type does not match
+            that specified by `vartype` (or `ens_validate`).
+        vardoc : str, optional
```
Comment (scopatz) [inline xonsh/environ.py]: this could be called doc

File: xonsh/environ.py [hunk 3]
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
...
+                 doc_configurable=None, doc_default=None, doc_store_as_str=None):
+        """Register an enviornment variable with optional type handling,
+        default value, doc.
+
+        Parameters
+        ----------
+        varname : str
+            Environment variable name to register. Typically all caps.
+        vartype : str, optional,  {'bool', 'str', 'path', 'int', 'float'}
+            Variable type. If not one of the available presets, use `ens_validate`,
```
Comment (scopatz) [inline xonsh/environ.py]: These could probably drop `var`

File: xonsh/environ.py [hunk 4]
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
...
+            `ens_convert`, and `ens_detype` to specify type behavior.
+        defaultval : optional
```
Comment (scopatz) [inline xonsh/environ.py]: This could probably just be `default`

File: xonsh/environ.py [hunk 5]
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
...
+        # set default value for envvar
+        if defaultval is not None:
+            if ensurer['validate'](defaultval):
+                self._defaults[varname] = defaultval
+            else:
+                raise ValueError("Default value does not match type specified by ensurer")
+
+        # set doc for envvar
+        if vardoc is not None:
+            self._docs[varname] = VarDocs(
```
Comment (scopatz) [inline xonsh/environ.py]: Ideally I think we'd combine VarDocs and Ensurer into a single namedtuple

**Comment by scopatz** [comment]
Thanks for putting this in @dotsdl! Sorry I have had a couple of crazy weeks and didn't take a look sooner

File: xonsh/environ.py (hunk 1 above)
Comment (dotsdl) [reply xonsh/environ.py]: Done!

File: xonsh/environ.py (hunk 2 above)
Comment (dotsdl) [reply xonsh/environ.py]: Done!

File: xonsh/environ.py (hunk 3 above)
Comment (dotsdl) [reply xonsh/environ.py]: Done!

File: xonsh/environ.py (hunk 4 above)
Comment (dotsdl) [reply xonsh/environ.py]: Done!

File: xonsh/environ.py (hunk 5 above)
Comment (dotsdl) [reply xonsh/environ.py]: Can do. Do we want to merge `DEFAULT_ENSURERS` and `DEFAULT_DOCS` along with this change (seems inevitable). Would we also want to merge in `DEFAULT_VALUES` to this new namedtuple?

Comment (dotsdl) [reply xonsh/environ.py]: ping @scopatz; do we want to merge together all the `DEFAULT_*` lazy objects?

**Comment by gforsyth** [comment]
Hey @dotsdl !  Sorry for the radio silence -- what's the status on this?  I'm seeing `black` failures that should be pretty quick to fix, is anything else needed here?

**Comment by dotsdl** [comment]
@gforsyth there is one unresolved question for @scopatz up above. I am happy to implement once I have clarity on what the desired end state is.

**Comment by astronouth7303** [comment]
@dotsdl @scopatz ping

File: xonsh/environ.py (hunk 5 above)
Comment (scopatz) [reply xonsh/environ.py]: Yes, I think we do.  I am so sorry for taking so long to get back to this here

Comment (dotsdl) [reply xonsh/environ.py]: No worries, all good! I will proceed with implementation. I believe I can carve out some time next week.

Comment (scopatz) [reply xonsh/environ.py]: Awesome! thanks @dotsdl!

**Comment by dotsdl** [comment]
I believe I've got everything needed in place. Now need to address tests.

**Comment by dotsdl** [comment]
I'm getting test failures that I seriously don't understand. If anyone has time to glance, I'll take any ideas you have for what might be happening here.

**Comment by scopatz** [comment]
I think that the issue here is that the history replayer didn't get updated to use the new API. We could consider getting rid of it, since I am not sure anyone uses it.

**Comment by gforsyth** [comment]
Hey @dotsdl -- let's just remove history replay.

**Comment by dotsdl** [comment]
I believe this is ready to go. I'm not sure what's going on with the CI, but all tests pass for me locally here. This one has been on my list to revisit for a long time, so thank you for everyone's patience.

I've removed `replay.py` and `test_replay.py`, so no more history replay. I've also added some explicit tests for registration/deregistration of env variables. Please check that these tests align with your expectations for how this should work.

Anything additional needed for merge?

**Comment by scopatz** [comment]
@xonsh/xore - this is ready for review! I approve of @dotsdl's changes, but I did some cleanup here, so if someone else wants to look at it, that would be good too

**Comment by dotsdl** [comment]
Awesome, thank you @scopatz! I saw the flurry of activity and got excited. Thanks for helping me get it to the finish line!

**Comment by scopatz** [comment]
Of course!

**Comment by gforsyth** [comment]
:rocket: thanks @dotsdl and @scopatz !!

**Comment by laloch** [comment]
Huh, really? I completely missed this one. This removes ensurers!

**Comment by scopatz** [comment]
It changes the API so that so can specify EnvVar behavior in one place.

**Comment by laloch** [comment]
Yes, I understand its purpose, but it also removes the possibility to restrict a variable to a programmable set of values.

**Comment by scopatz** [comment]
It shouldn't.  The ensurers are still there.

**Comment by laloch** [comment]
Oh, now I see! You're right. Sorry for the noise.

**Comment by scopatz** [comment]
What do you think is missing @laloch?

**Comment by laloch** [comment]
Everything is in place. I'm an idiot. Sorry. :facepalm:

**Comment by scopatz** [comment]
:+1: Yay! I was scared for a moment!

**Pull request merged**